### ISSUE TITLE
Punitive Counterstrike text shows old and new trace formatting on top of each other

### ISSUE THREAD

**Issue by infinitenutshell** [Null-Signal-Games/netrunnerdb#623]
https://netrunnerdb.com/en/card/31078
New trace style (expected): Trace[5]. If successful...
Old trace style (still seen on cards that haven't been reprinted or added to the Card Text Updates document): **Trace [5] –** If successful...
Punitive currently displays as: **Trace [5] –**. If successful...

Probably there's just an extraneous bit of markup in the card data. This error does not currently appear for any other card.

Edit: fixed en dashes

**Comment by plural** [comment]
Ah, interesting.

This isn't a problem with the card data.  This looks to be a rare instance (only?) of a Trace effect just being a full sentence.

The controller currently replaces `<trace>Trace 5</trace>` like this:
` $cardinfo['text'] = preg_replace('/<trace>([^<]+) ([X\d]+)<\/trace>/', '<strong>\1 [\2]</strong>–', $cardinfo['text']);`

To fix this, we would actually need to update all other traces in the JSON to add the en dash as a literal and then remove that en dash from the controller.

**Comment by infinitenutshell** [comment]
I think you're overcomplicating it a bit. All you need to do is remove the `<trace>` tag that applies the old formatting. All these cards display the new formatting just fine: https://netrunnerdb.com/find/?q=x%3A%22trace%5B%22&sort=name&view=text&_locale=en

**Comment by plural** [comment]
Those don't actually use the trace syntax.  They just have a literal Trace[#] in the text, like Executive Function.
Compare:
`    "text": "[subroutine] Trace[4]. If successful, do 1 brain damage.",`

to CtM's: 
` "text": "The first time the Runner trashes an installed Corp card each turn, you may <trace>trace 4</trace> If successful, give the Runner 1 tag (cannot be avoided).",`

**Comment by plural** [comment]
incidentally, this also makes them normal weight and not bold.

**Comment by plural** [comment]
if we have a trace syntax, why use it sometimes, but not other times?  Is trace a bit of text or an ability with a keyword/formatting?  Right now it's 🤷   It would be better for data quality to have it be a real keyword so we can match on the ability. The text formatting is pretty inconsistent now with respect to what is a keyword and what is some literal business shoved in (hello html tags grumble grumble).

**Comment by infinitenutshell** [comment]
I noticed today that this issue now affects all trace cards updated in CTU1.3. Those cards are:

Power Grid Overload
Data Hound
Midseason Replacements
Hellion Alpha Test
Information Overload
Archangel
Turnpike
Waiver
Hard-Hitting News
Hellion Beta Test
IP Block
Thoth
Forced Connection
Trojan Horse
Kill Switch
Data Raven
SEA Source
and Punitive Counterstrike is still affected as well.

**Comment by infinitenutshell** [comment]
> Those don't actually use the trace syntax. They just have a literal Trace[#] in the text, like Executive Function. Compare: ` "text": "[subroutine] Trace[4]. If successful, do 1 brain damage.",`
> 
> to CtM's: ` "text": "The first time the Runner trashes an installed Corp card each turn, you may <trace>trace 4</trace> If successful, give the Runner 1 tag (cannot be avoided).",`

All the cards listed in my previous comment should match the Executive Functioning style. Normal weight is correct.

For reference, here are all cards that have correct new-style trace formatting:

Sherlock 1.0
Ichi 2.0
Viktor 2.0
Self-destruct
Vikram 1.0
Sherlock 2.0
SYNC BRE
Warroid Tracker
Threat Level Alpha
Executive Functioning
Holmegaard
Black Hat
Ichi 1.0

And here are all cards that still use old trace formatting, and this is expected because they haven't yet appeared in a CTU release:

Matrix Analyzer
Restructured Datapool
TMI
Draco
Viper
Uroboros
Salvage
Foxfire
Burke Bugs
Invasion of Privacy
Muckraker
Snoop
Broadcast Square
Shinobi
Taurus
Manhunt
Sagittarius
Gemini
Snatch and Grab
Troll
Virgo
Checkpoint
Searchlight
Gutenberg
News Hound
Resistor
Assassin
Security Nexus
Another Day, Another Paycheck
NBN: Controlling the Message
Enforcing Loyalty
Hatchet Job
Door to Door
Citadel Sanctuary
Drone Screen
Veritas
Macrophage
The Archivist
Traffic Analyzer
Bernice Mai
Shadow
No One Home
White Hat
Kuwinda K4H1U3
Tempus
Amani Senai
Intake
Surveyor
SIU
Eavesdrop
Ash 2X3ZB9CY
Flare
Caduceus
Paper Trail
Hunter
Trebuchet
Scapenet
Winchester

**Comment by infinitenutshell** [comment]
> if we have a trace syntax, why use it sometimes, but not other times? Is trace a bit of text or an ability with a keyword/formatting? Right now it's 🤷 It would be better for data quality to have it be a real keyword so we can match on the ability. The text formatting is pretty inconsistent now with respect to what is a keyword and what is some literal business shoved in (hello html tags grumble grumble).

Yeah, it's inconsistent. The card pool as a whole is inconsistent in this and several other ways, because the Rules team does not have bandwidth to update card text for the entire backlog at once. We'll get there eventually, and then the old trace formatting will be fully obsolete. If you want to be done with the HTML tag before then, I have no objection to replacing e.g. `<trace>trace 4</trace>` with `Trace[4] - `  for all cards that have not been updated.

Actually maybe I'm misunderstanding what you ultimately want. What do you mean by a "real keyword"? Would eliminating the `<trace>` tag entirely be a good thing or a bad thing from your point of view?

**Comment by mankinchi** [comment]
I will look into this one. This is my personal preference:
* changing all the cards not using `<trace>` tag to use them for consistency
* update the `<trace>` card format to fit the new styling

In my opinion, keep the `<trace>` tag is much better in term of searching down the road since we can use it to easily identify the `<trace>` keyword with just a normal random `trace` word in the paragraph. What do you guys think?

**Comment by mankinchi** [comment]
also, @plural, I'm seeing that we are doing `<trace>Trace 5</trace>`. If this is meant to use for `<trace>` only, would you mind if I change the format to just `<trace>5</trace>` instead? I can make the code still work with any cards that have the old format in it.

**Comment by infinitenutshell** [comment]
I suppose there is a case to be made for being able to search cards that _perform_ a trace separately from cards that _mention_ trace. Not that we're going to have that functionality until NRDB2 regardless. How does scryfall implement "keyword:" searches? Netrunner is a bit trickier than MtG because most of our keywords are just words and phrases that we use in sentences, not self-contained abilities consisting of just a word or two. I'll defer to y'all in the trenches as to whether it's better to have additional markup in the text field or have extra fields. 

I do think it could create weirdness down the road to have the internal representation diverge from the actual text, as the previous comment proposes. Would having just the `<trace>` tag without the non-tag occurence of `trace` in the text mess up search strings? Suppose for example I want to find cards that can optionally perform a trace. My first thought for the search to use would be `x:"may trace"`.

**Comment by mankinchi** [comment]
Pretext: I'm totally new to netrunner so there r so many cards that I'm not aware of.

My goal is to distinguish between card effects like this `when you trace, you might do abc` with `Trace [5]. If successful, do something`. So if you want to search for something with the `trace` keyword, then `x: trace` would work perfectly. I'm not sure whether we have something like this implemented, but I'm assuming we can do something like `a:trace` (a being action, for example) to search for cards that allow you to have the action of `trace`. This is the main reason why I think having a specific trace tag would be a good choice.

**Comment by infinitenutshell** [comment]
Welcome!

Practically speaking, there are only a handful of cards with "when you trace" and similar effects, and NSG has moved away from actively using the trace mechanic to begin with, so there aren't likely to be more of those cards any time soon. I do find myself searching for all kinds of weird things for Rules team work, so I would probably get _some_ use out of a specialized functionality like you are suggesting. But it's also pretty easy to search `x:trace x:"If successful"|"If unsuccessful"`, so maybe it isn't worth the implementation effort? I recommend following @plural's lead here

**Comment by mankinchi** [comment]
at this point, for the scope of this issue, there won't be any implementation for the searching one. I just put that out there to prove my point of having a `<trace> tag should benefit us much better than leaving them as normal text. 

The code change for this style change is quite minimal and I have already done it. I'm working on some data changes to switch the plain text of `Trace[x]` to use `<trace>` tag instead.
